strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,91 @@ def context(self, *args, copies=1, shift="default"):
+            )
+        # Second check: the gates act in the correct modes
+        program_modes = [[r.ind for r in cmd.reg] for cmd in self.rolled_circuit]
+        device_modes = [op["modes"] for op in device_layout.operations]
+        if program_modes != device_modes:
+            raise CircuitError(
+                "Program cannot be used with the device '{}' "
+                "due to incompatible mode ordering.".format(device.target)
+            )
+        # Third check: the parameters of the gates are valid
+        gate_params_ranges = device.gate_parameters
+        # We will loop over the different operations in the device specification
+        for i, operation in enumerate(device_layout.operations):
+            # We obtain the name of the parameter(s)
+            param_names = operation["args"]
+            len_params_program = len(self.rolled_circuit[i].op.p)
+            len_params_device = len(param_names)
+            # The next if is to make sure we do not flag incorrectly things like Sgate(r,0) beign different Sgate(r)
+            if len_params_device < len_params_program:
+                for j in range(1, len_params_program):
+                    if self.rolled_circuit[i].op.p[j] != 0:
+                        raise CircuitError(
+                            "Program cannot be used with the device '{}' "
+                            "due to incompatible parameter.".format(device.target)
+                        )
+            # Now we will check explicitly if the parameters in the program match
+            counter = 0  # counts the number of symbolic variables, which are labeled consecutively by the context method
+            for k, param_name in enumerate(param_names):
+                # Obtain the relelvant parameter range from the device
+                param_range = device.gate_parameters[param_name]
+                # Obtain the value of the corresponding parameter in the program
+                program_param = self.rolled_circuit[i].op.p[k]
+                if sf.parameters.par_is_symbolic(program_param):
+                    # If it is a symbolic value go and lookup its corresponding list in self.tdm_params
+                    local_p_vals = self.tdm_params[counter]
+                    print(len(local_p_vals), device.modes["temporal"]["max"])
+                    if len(local_p_vals) > device.modes["temporal"]["max"]:
+                        raise CircuitError(
+                            "Program cannot be used with the device '{}' "
+                            "due to not having enough temporal modes.".format(device.target)
+                        )
+
+                    for x in local_p_vals:
+                        if not x in param_range:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter.".format(device.target)
+                            )
```

Comment:
I guess this could include the parameter range and the value as well. :slightly_smiling_face: